pkg/controller/util/endpoint/controller_utils.go

```diff
@@ -32,6 +34,75 @@ import (
 // EndpointsMatch is a type of function that returns true if pod endpoints match.
 type EndpointsMatch func(*v1.Pod, *v1.Pod) bool
 
+// ServiceSelectorCache is a cache of service selectors to avoid high CPU consumption caused by frequent calls to AsSelectorPreValidated (see #73527)
+type ServiceSelectorCache struct {
+	cache map[string]labels.Selector
+	lock  sync.Mutex
+}
+
+// NewServiceSelectorCache init ServiceSelectorCache for both endpoint controller and endpointSlice controller.
+func NewServiceSelectorCache() *ServiceSelectorCache {
+	return &ServiceSelectorCache{
+		cache: map[string]labels.Selector{},
+	}
+}
+
+// Get return selector and existence in ServiceSelectorCache by key.
+func (sc *ServiceSelectorCache) Get(key string) (labels.Selector, bool) {
+	selector, ok := sc.cache[key]
+	return selector, ok
+}
+
+// Update can update or add a selector in ServiceSelectorCache while service's selector changed.
+func (sc *ServiceSelectorCache) Update(key string, rawSelector map[string]string) labels.Selector {
+	sc.lock.Lock()
+	defer sc.lock.Unlock()
+	selector := labels.Set(rawSelector).AsSelectorPreValidated()
+	sc.cache[key] = selector
+	return selector
+}
+
+// Delete can delete selector which exist in ServiceSelectorCache.
+func (sc *ServiceSelectorCache) Delete(key string) {
+	sc.lock.Lock()
+	defer sc.lock.Unlock()
+	delete(sc.cache, key)
+}
+
+// GetPodServiceMemberships returns a set of Service keys for Services that have
+// a selector matching the given pod.
+func (sc *ServiceSelectorCache) GetPodServiceMemberships(serviceLister v1listers.ServiceLister, pod *v1.Pod) (sets.String, error) {
+	set := sets.String{}
+	services, err := serviceLister.Services(pod.Namespace).List(labels.Everything())
+	if err != nil {
+		// don't log this error because this function makes pointless
+		// errors when no services match
+		return set, nil
+	}
```

Comment:
This could end up hiding significant bugs. As far as I can tell, the original implementation of this did return an error here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/listers/core/v1/service_expansion.go#L37-L40. Could we just continue to return this error here as the original implementation seems to? Is there a specific error you can catch specifically if no services match?

Reply:
these sentences come from here: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/util/endpoint/controller_utils.go#L87-L90, and the error is a return of service list: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/listers/core/v1/service_expansion.go#L37-L40
I don't know whether any exception happens return set, err rather than set, nil

Reply:
After review it carefully, it seems should return set, err (original code may be wrong)? After all, I will address it as `return set, err`